wasm/src/strings.c

```diff
@@ -404,6 +405,22 @@ opa_value *opa_strings_split(opa_value *a, opa_value *b)
     return &arr->hdr;
 }
 
+opa_value *opa_strings_sprintf(int32_t id, void *ctx, opa_value *a, opa_value *b)
+{
+    if (opa_value_type(a) != OPA_STRING || opa_value_type(b) != OPA_ARRAY)
+    {
+        return NULL;
+    }
+
+    opa_string_t *s = opa_cast_string(a);
+    opa_array_t *arr = opa_cast_array(b);
+
+    // TODO: Implement basic sprintf support but revert to
+    // opa_builtin2, if not capable of the requested formatting.
+
+    return opa_builtin2(id, ctx, a, b);
+}
```

Comment:
The current approach seems like a performance optimization because the host environment has no way to know whether the sprintf function needs to be provided. For correctness it would always have to provide an implementation.

An alternative approach would be to only mark the built-in function as external if the partial implementation did not support it. If the analysis could be performed in the wasm compiler then we could emit wasm executables that do not have any dependency on host implementations of partial functions. This would be nice from a portability perspective (which would make authoring and management easier IMO). This is what I'd originally thought of doing. However...

I'm not sure how easy it would be to perform the analysis in the wasm compiler--the analysis is likely the same as what the C library would have to do anyway so duplicating that code in Go seems wasteful. If we could invoke the C library from the wasm compiler that would avoid it, but that comes with it's own downsides.